boa_engine/src/builtins/intl/date_time_format.rs

```diff
@@ -234,3 +259,1208 @@ pub(crate) fn to_date_time_options(
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma402/#sec-internal-slots
+fn build_available_locales() -> Vec<JsString> {
+    let default_locale_str = default_locale();
+    let canonicalized_locale = default_locale_str.replace('_', "-");
+    let splitted_locale: Vec<&str> = canonicalized_locale.split('-').collect();
+    let default_locale_fallback = splitted_locale
+        .get(0)
+        .expect("Failed to split default locale");
+    let available_locales = vec![default_locale_str, JsString::new(default_locale_fallback)];
+
+    available_locales
+}
+
+/// The `DefaultTimeZone` abstract operation returns a String value representing the valid and
+/// canonicalized time zone name for the host environment's current time zone.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma402/#sec-defaulttimezone
+fn default_time_zone() -> JsString {
+    // FIXME fetch default time zone from the environment
+    JsString::new("UTC")
+}
+
+/// The abstract operation `IsValidTimeZoneName` takes argument `timeZone`, a String value, and
+/// verifies that it represents a valid Zone or Link name of the IANA Time Zone Database.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma402/#sec-isvalidtimezonename
+pub(crate) fn is_valid_time_zone_name(time_zone: &JsString) -> bool {
+    let time_zone_str = time_zone.to_string();
+    let maybe_time_zone: Result<Tz, _> = time_zone_str.parse();
+    maybe_time_zone.is_ok()
+}
+
+/// The abstract operation `CanonicalizeTimeZoneName` takes argument `timeZone` (a String value
+/// that is a valid time zone name as verified by `IsValidTimeZoneName`). It returns the canonical
+/// and case-regularized form of timeZone.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma402/#sec-canonicalizetimezonename
+pub(crate) fn canonicalize_time_zone_name(time_zone: &JsString) -> JsString {
+    let time_zone_str = time_zone.to_string();
+    let canonical_tz: Tz = time_zone_str
+        .parse()
+        .expect("CanonicalizeTimeZoneName: time zone name is not supported");
+    JsString::from(canonical_tz.name())
+}
+
+/// Converts `hour_cycle_str` to `preferences::HourCycle`
+pub(crate) fn string_to_hour_cycle(hour_cycle_str: &JsString) -> preferences::HourCycle {
+    let mut string_to_hc_map = FxHashMap::<JsString, preferences::HourCycle>::default();
+    string_to_hc_map.insert(JsString::from("h11"), preferences::HourCycle::H11);
+    string_to_hc_map.insert(JsString::from("h12"), preferences::HourCycle::H12);
+    string_to_hc_map.insert(JsString::from("h23"), preferences::HourCycle::H23);
+    string_to_hc_map.insert(JsString::from("h24"), preferences::HourCycle::H24);
+
+    let hour_cycle = string_to_hc_map
+        .get(hour_cycle_str)
+        .expect("Invalid hour cycle");
+    *hour_cycle
+}
```

Comment:
Using a `match` on `hour_cycle_str.as_str()` here is way more efficient than the allocations of the map and strings.